lib/solarwinds_apm/support/resource_detector/aws/ecs.rb

```diff
@@ -0,0 +1,156 @@
+# frozen_string_literal: true
+
+# © 2023 SolarWinds Worldwide, LLC. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at:http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+
+require 'net/http'
+require 'uri'
+require 'json'
+require 'socket'
+
+module SolarWindsAPM
+  module ResourceDetector
+    module ECS
+      module_function
+
+      CONTAINER_ID_LENGTH = 64
+      DEFAULT_CGROUP_PATH = '/proc/self/cgroup'
+      HTTP_TIMEOUT = 1
+
+      def detect
+        ecs_instance = ENV['ECS_CONTAINER_METADATA_URI_V4'] || ENV.fetch('ECS_CONTAINER_METADATA_URI', nil)
+        unless ecs_instance
+          SolarWindsAPM.logger.warn 'AwsEcsDetector: Process is not on ECS'
+          return ::OpenTelemetry::SDK::Resources::Resource.create({})
+        end
+
+        gather_data
+      end
+
+      # curl http://169.254.169.254/latest/meta-data/hostname
+      def gather_data
+        attribute = {
+          ::OpenTelemetry::SemanticConventions::Resource::CLOUD_PROVIDER => 'aws',
+          ::OpenTelemetry::SemanticConventions::Resource::CLOUD_PLATFORM => 'aws_ecs',
+          ::OpenTelemetry::SemanticConventions::Resource::CONTAINER_ID => resolve_container_id,
+          ::OpenTelemetry::SemanticConventions::Resource::HOST_NAME => Socket.gethostname
+        }
+
+        metadata_url = ENV.fetch('ECS_CONTAINER_METADATA_URI_V4', nil)
+        if metadata_url
+          container_metadata = get_url_as_json(metadata_url)
+          task_metadata = get_url_as_json("#{metadata_url}/task")
+
+          merge_metadata(attribute, container_metadata, task_metadata)
+        end
+
+        attribute.compact!
+        ::OpenTelemetry::SDK::Resources::Resource.create(attribute)
+      end
+
+      def resolve_container_id
+        container_id = nil
+        begin
+          raw_data = File.read(DEFAULT_CGROUP_PATH).strip
+          raw_data.each_line do |line|
+            if line.length > CONTAINER_ID_LENGTH
+              container_id = line[-CONTAINER_ID_LENGTH..]
+              break
+            end
+          end
+        rescue StandardError => e
+          SolarWindsAPM.logger.warn("AwsEcsDetector failed to read container ID: #{e}")
```

Comment:
Should be debug level log here and the code below?

Reply:
Updated (also other warn in resource_detector/aws)